docs/src/examples/ITensor.md

```diff
@@ -17,6 +17,165 @@ or this
 println("T inds = ",inds(T))
 ```
 
+## Getting and Setting Elements of an ITensor
+
+Say we have an ITensor constructed as:
+
+```julia
+i = Index(3,"index_i")
+j = Index(2,"index_j")
+k = Index(4,"index_k")
+
+T = ITensor(i,j,k)
+```
+
+An ITensor constructed this way starts with all of its elements
+equal to zero. (Technically it allocates no storage at all but this is
+an implementation detail.)
+
+**Setting Elements**
+
+To set an element of this ITensor, such as the element where `(i,j,k) = (2,1,3)`,
+you can do the following:
+
+```julia
+T[i=>2,j=>1,k=>3] = -3.2
+```
+
+In the Julia language, the notation `a=>b` is a built-in notation for making a `Pair(a,b)`
+object.
+
+Because the Index objects are passed to `T` along with their values, passing them in a different order has exactly the same effect:
+
+```julia
+# Both of these lines of code do the same thing:
+T[j=>1,i=>2,k=>3] = -3.2
+T[j=>1,k=>3,i=>2] = -3.2
+```
+
+**Getting Elements**
+
+You can retrieve individual elements of an ITensor by accessing them through the same notation used to set elements:
+
+```julia
+el = T[j=>1,i=>2,k=>3]
+println("The (i,j,k) = (2,1,3) element of T is ",el)
+```
+
+## Arithmetic With ITensors
+
+ITensors can be added and subtracted and multiplied by scalars just like plain tensors can. But ITensors have the additional feature that you can add and subtract them even if their indices are in a different order from each other, as long as they have the same collection of indices.
+
+For example, say we have ITensors `A`, `B`, and `C`:
+```julia
+i = Index(3,"i")
+j = Index(2,"j")
+k = Index(4,"k")
+
+A = randomITensor(i,j,k)
+B = randomITensor(i,j,k)
+C = randomITensor(k,i,j)
+```
+Above we have initialized these ITensors to have random elements, just for the sake of this example.
+
+We can then add or subtract these ITensors
+
+```julia
+R1 = A+B
+R2 = A-B
+R3 = A+B-C
+```
+
+or do more complicated operations involving real and complex scalars too:
+
+```julia
+R4 = 2.0*A - B + C/(1+1im)
+```
+
+## Elementwise Operations on ITensors
+
+ITensor objects support Julia broadcasting operations, making it quite easy to carry out element-wise operations on them in a very similar way as for regular Julia arrays. As a concrete example, consider the following ITensor initialized with random elements
+
+```julia
+i = Index(2,"i")
+j = Index(3,"j")
+
+T = randomITensor(i,j)
+```
+
+Here are some examples of basic element-wise operations we can do using Julia's dotted operator broadcasting syntax.
+
+```julia
+# Multiply every element of `T` by 2.0:
+T .*= 2.0
+```
+
+```julia
+# Add 1.5 to every element of T
+T .+= 1.5
```

Comment:
I'm not sure this will do the correct thing for QN ITensors unfortunately:
```julia
i = Index([QN(0) => 1, QN(1) => 1])
A = emptyITensor(i', dag(i))
A[1, 1] = 1
A .+= 1.5
@show A
```
Gives:
```julia
A = ITensor ord=2
Dim 1: (dim=2|id=571)' <Out>
 1: QN(0) => 1
 2: QN(1) => 1
Dim 2: (dim=2|id=571) <In>
 1: QN(0) => 1
 2: QN(1) => 1
NDTensors.BlockSparse{Float64, Vector{Float64}, 2}
 2×2
Block(1, 1)
 [1:1, 1:1]
 2.5

Block(2, 1)
 [2:2, 1:1]
 1.5

Block(1, 2)
 [1:1, 2:2]
 1.5

Block(2, 2)
 [2:2, 2:2]
 1.5
```
I'm also not even sure how well defined this operation is for QN ITensors, should it only add to the existing QN blocks, any possible nonzero flux block, etc? I guess for now we should just disallow it for QN ITensors.

Reply:
Hm, yes I see. I should slow down on the larger docs PR's and make sure to leave time for review. Thanks for taking a look!

Yes I think for that operation it's best for now to disallow it for QN ITensors. Maybe we could separately make a rule that if you do any function of the form `T .= f.(T)` then it will be carried out on all elements which are compatible with the flux (so it would introduce any missing, but allowed, blocks) but still not act on any disallowed blocks. Just an idea.

Reply:
Yeah, my issue with that is it doesn't really feel mathematically correct based on the definition of broadcasting being an elementwise operation on all elements of a structure, and the way it generally is defined for sparse structures:
```julia
julia> using LinearAlgebra

julia> D = Diagonal(randn(4))
4×4 Diagonal{Float64, Vector{Float64}}:
 -1.28286   ⋅         ⋅         ⋅ 
   ⋅       0.15365    ⋅         ⋅ 
   ⋅        ⋅       -1.38377    ⋅ 
   ⋅        ⋅         ⋅       -1.87302

julia> D .+= 2.4
ERROR: ArgumentError: cannot set off-diagonal entry (2, 1) to a nonzero value (2.4)
```
I don't think that is a common enough operation to do to warrant making such a convenient and potentially confusing operation. I think for that usage we should provide separate functionality for that, like:
```julia
insert_zero_blocks!(A) # Insert blocks of zero consistent with the flux
nonzeros(A) .+= 1.5  # Now broadcast only on the nonzero values of A
```
Or alternative, you construct another QN ITensor `B` which is filled with the value `1.5` and add it in-place to `A`, which may be just as efficient as the above. Another reason for not making the above operations default behavior for broadcasting is that in general the functionality `insert_zero_blocks!(A)` can be very slow, so now we only do it when necessary, like in constructors. (Note that `insert_zero_blocks!(A)` doesn't actually exist in that form but that kind of operation is used in QN constructors in a different form.)

Reply:
Ok agreed that's the best and clearest approach. It lets you do the thing I was saying, but then see from the code that it exactly what is happening and not have to guess. And it lets you fine-tune the behavior, or we could introduce other similar behaviors and wrappers in the future.